File: 8.Database-with-backend/code/typescript/isolation_the_lost_update_race_revisited_1.ts
import { Pool } from 'pg';

export async function transferFunds(pool: Pool, a: string, b: string): Promise<void> {
    const client = await pool.connect();
    
    try {
        await client.query('BEGIN');
        
        // safety net happens in catch block in TS/JS instead of defer
        
        await client.query('UPDATE accounts SET balance = balance - 500 WHERE id = $1', [a]);
        // if this throws -> catch block fires -> A is untouched
        
        await client.query('UPDATE accounts SET balance = balance + 500 WHERE id = $1', [b]);
        // if this throws -> catch block fires -> both reverted
        
        await client.query('COMMIT'); // only here do both writes become permanent
    } catch (err) {
        await client.query('ROLLBACK');
        throw err;
    } finally {
        client.release();
    }
}

File: 8.Database-with-backend/code/rust/isolation_the_lost_update_race_revisited_1.rs
use sqlx::{PgPool, Error};

pub async fn transfer_funds(pool: &PgPool, a: &str, b: &str) -> Result<(), Error> {
    let mut tx = pool.begin().await?;
    
    // In sqlx, transactions rollback automatically if dropped before commit
    // so it acts as our safety net implicitly!
    
    sqlx::query!("UPDATE accounts SET balance = balance - 500 WHERE id = $1", a)
        .execute(&mut *tx).await?;
    // if error occurs -> early return -> tx dropped -> A is untouched
        
    sqlx::query!("UPDATE accounts SET balance = balance + 500 WHERE id = $1", b)
        .execute(&mut *tx).await?;
    // if error occurs -> early return -> tx dropped -> both reverted
        
    tx.commit().await?; // only here do both writes become permanent
    Ok(())
}
